DOC: Fixes for 18 broken links

bjnath opened the pull request:
This, with PR #16465, should fix nearly all the remaining broken links
on the site. 4 or 5 others should be easy to fix and just
need attention from someone more knowledgeable -- will
open an issue. For release notes with dead links,
I could usually find links on archive.org for roughly contemporary
versions.

bjnath committed:
DOC: Fixes for 18 broken links

This, with PR #16465, should fix nearly all the remaining broken links
on the site. 4 or 5 others should be easy to fix and just
need attention from someone more knowledgeable -- will
open an issue. For release notes with dead links,
I could usually find links on archive.org for roughly contemporary
versions.

charris commented on doc/HOWTO_RELEASE.rst.txt:
```diff
@@ -19,18 +19,20 @@ Source tree
 - https://github.com/numpy/numpy/blob/master/doc/HOWTO_RELEASE.rst.txt
-- http://projects.scipy.org/numpy/wiki/MicrosoftToolchainSupport (dead link)
+
+.. - http://projects.scipy.org/numpy/wiki/MicrosoftToolchainSupport (dead link)
```
Should "dead link" still be there?

bjnath replied:
Sorry, @charris, I meant that to be commented out as well.  

What I *wanted* to do was leave both the link and the annotation completely alone. If I knew a way of "escaping" a URL so the tools would treat it like a plain string, that's what I would have done.

charris replied:
I hadn't seen that before, does `..` actually work to make it a comment?

seberg replied:
We could provide a historical link for now: https://web.archive.org/web/20130402024847/http://projects.scipy.org/numpy/wiki/MicrosoftToolchainSupport (not sure if the 2012 version is not too old).

eric-wieser replied:
Using `\http://projects.scipy.org/numpy/wiki/MicrosoftToolchainSupport` seems to make the link non-clickable for me, but an archive link is probably more useful anyway.

bjnath replied:
Do these links have current relevance? I didn't look in archive.org because I was hesitant to resurrect links for a process that no longer exists.

bjnath replied:
> does .. actually work to make it a comment?

I learned about it from 
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#comments

seberg replied:
Oh... I suppose just commenting it out or even deleting may be fine as well here. I guess @charris was only wondering if it worked to hide/remove the link, which the `.. ` directive does...

bjnath replied:
> wondering if it worked to hide/remove the link
Oh, I see. Looking at the docs it really could go either way; sorry @charris to misunderstand you.

> deleting may be fine
They were historic. Now they're history.

eric-wieser commented on doc/source/reference/arrays.classes.rst:
```diff
@@ -173,7 +173,7 @@ NumPy provides several hooks that classes can customize:
 
    -  ``func`` is an arbitrary callable exposed by NumPy's public API,
       which was called in the form ``func(*args, **kwargs)``.
-   -  ``types`` is a `collection <collections.abc.Collection>`_
```
This ought to be handled by intersphinx.

eric-wieser replied:
Perhaps `python:collections.abc.Collection`, or `` :class:`collection <collections.abc.Collection>` `` is needed here.

bjnath replied:
```
-  ``types`` is a collection :py:class:`collections.abc.Collection`
```
Thanks for the suggestion, putting it in.

eric-wieser commented on doc/source/reference/global_state.rst:
```diff
@@ -82,4 +82,4 @@ debug option can be interesting for testing code written
 in C which iterates through arrays that may or may not be
 contiguous in memory.
 Most users will have no reason to change these, for details
-please see the `memory layout <memory-layout>`_ documentation.
```
This label exists. Perhaps this needs `:ref:`, or something has gone wrong with numpy-doc?

https://github.com/numpy/numpy/blob/3e261be81a5f2645ffc4c1e79a5dccc65a216669/doc/source/reference/arrays.ndarray.rst#L90-L93

bjnath replied:
Thanks, will change:
```
see the :ref:`memory layout <memory-layout>` documentation.
```

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1974,7 +1974,7 @@ class vectorize:
     ----------
     .. [1] NumPy Reference, section `Generalized Universal Function API
-           <https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html>`_.
+           <https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html>`_.
```
This feels like we should be using an internal link.

bjnath replied:
I'm surprised not to see this case covered in the [docstring guide](https://numpy.org/devdocs/docs/howto_document.html#numpydoc-docstring-guide).

My thought was that content of the .py files should be self-contained.  

If we link internally I'll need a markup suggestion.

bjnath committed:
DOC: Update to "Fixes for 18 broken links #16472"

* Obsolete links, previously commented out, now deleted:
https://github.com/numpy/numpy/pull/16472#discussion_r433928958

* Semantic markup for reference to Python class:
https://github.com/numpy/numpy/pull/16472#discussion_r433553928

* Missing :ref: in internal link:
https://github.com/numpy/numpy/pull/16472#discussion_r433554484

Not included: Resolution on using external/internal doc link in .py:
https://github.com/numpy/numpy/pull/16472#discussion_r433554824

bjnath commented:
b9369af updates the PR for all comments but the last (internal/external link to numpy in docstring).

rossbar commented on numpy/lib/function_base.py:
```diff
@@ -1974,7 +1974,7 @@ class vectorize:
     References
     ----------
     .. [1] NumPy Reference, section `Generalized Universal Function API
```
```suggestion
    .. [1] :doc:`/reference/c-api/generalized-ufuncs`
```
This should do the trick for an internal link - the url below can be deleted if the doc role is used here.

bjnath committed:
DOC: Add internal link for 'Fixes for 18 broken links' PR #16472

Making reference [1] an internal link in function_base.py => numpy.vectorize.html

bjnath commented:
5c08dad resolves all comments.

rossbar commented on doc/source/dev/governance/governance.rst:
```diff
@@ -381,14 +381,13 @@ A list of current Institutional Partners is maintained at the page
 ================
 
-https://github.com/numpy/numpy/commits/master/doc/source/dev/governance/governance.rst
+https://github.com/numpy/numpy/blob/master/doc/source/dev/governance/governance.rst
```
I think the original link was actually correct here, at least given the section title - it seems that showing the commit history (rather than the current version) was intentional.

bjnath replied:
I agree. This was the one case in the PR where I changed a link that had not been reported bad.  The checker failed with 429's on github, so github checking had to be done by hand.  

Changing it back.

rossbar commented on doc/source/reference/random/new-or-different.rst:
```diff
@@ -15,7 +15,7 @@ What's New or Different
   `RandomState.standard_t`.
 
-Quick comparison of legacy `mtrand <legacy>`_ to the new `Generator`
+Quick comparison of legacy :ref:`mtrand <legacy>` to the new `Generator`
```
This link wasn't actually broken, but adding the explicit `:ref:` role doesn't hurt.

bjnath replied:
Please see my comment below: https://github.com/numpy/numpy/pull/16472#issuecomment-639450633

rossbar commented on doc/source/reference/ufuncs.rst:
```diff
@@ -109,7 +109,7 @@ The output of the ufunc (and its methods) is not necessarily an
 Indeed, if any input defines an :obj:`~class.__array_ufunc__` method,
 control will be passed completely to that function, i.e., the ufunc is
-`overridden <ufuncs.overrides>`_.
+`overridden <https://www.numpy.org/neps/nep-0013-ufunc-overrides.html>`_.
```
I think this was intended to be a link to the ["Overriding Ufunc Behavior"](https://numpy.org/devdocs/reference/ufuncs.html#overriding-ufunc-behavior) section of the same document, rather than an external reference to the NEP.

bjnath replied:
Agreed. Fixing original link.

rossbar commented on doc/source/release/1.19.0-notes.rst:
```diff
@@ -24,14 +24,14 @@ Calling ``np.array([[1, [1, 2, 3]])`` will issue a ``DeprecationWarning`` as
 warning.
 
-.. _`NEP 34`: https://numpy.org/neps/nep-0034.html
+.. _`NEP 34`: https://numpy.org/neps/nep-0034-infer-dtype-is-object.html
```
Now that #16450 is merged, the can be done via intersphinx rather than explicitly creating a link here.

mattip replied:
Are you sure? Check that is in `https://numpy.org/neps/object.inv` to `conf.py`

rossbar replied:
I see `nep34` as a valid target in the `objects.inv` file:

```
python -msphinx.ext.intersphinx https://numpy.org/neps/objects.inv
```

Good point though - it seems the neps are missing from the intersphinx mapping in the main `conf.py` for the NumPy docs. Adding it now...

bjnath replied:
Deferring NEP intersphinx enhancement for another PR.  This PR is for broken links.

rossbar commented on doc/source/user/quickstart.rst:
```diff
@@ -1054,7 +1056,7 @@ element is assumed to be the same along that dimension for the
 
 After application of the broadcasting rules, the sizes of all arrays
-must match. More details can be found in :doc:`basics.broadcasting`.
+must match. More details can be found in :ref:`basics.broadcasting`.
```
This link wasn't broken either - the `:doc:` role is preferred for linking to documents I think.

bjnath replied:
Please see my comment below: https://github.com/numpy/numpy/pull/16472#issuecomment-639450633

rossbar commented on numpy/lib/mixins.py:
```diff
@@ -68,7 +68,7 @@ class NDArrayOperatorsMixin:
     but that should support arithmetic and numpy universal functions like
     arrays as described in `A Mechanism for Overriding Ufuncs
-    <../../neps/nep-0013-ufunc-overrides.html>`_.
+    <https://numpy.org/neps/nep-0013-ufunc-overrides.html>`_.
```
Same comment as above - let's try intershpinx here

bjnath replied:
Deferring NEP intersphinx enhancement for another PR.  This PR is for broken links.

rossbar commented on numpy/polynomial/chebyshev.py:
```diff
@@ -104,7 +104,7 @@
 .. [1] A. T. Benjamin, et al., "Combinatorial Trigonometry with Chebyshev
   Polynomials," *Journal of Statistical Planning and Inference 14*, 2008
-  (preprint: https://www.math.hmc.edu/~benjamin/papers/CombTrig.pdf, pg. 4)
+  (https://web.archive.org/web/20080221202153/https://www.math.hmc.edu/~benjamin/papers/CombTrig.pdf, pg. 4)
```
There seems to be a non-preprint, non-web.archive link to this paper available from the original institution: https://math.hmc.edu/benjamin/wp-content/uploads/sites/5/2019/06/Combinatorially-Composing-Chebyshev-Polynomials.pdf

rossbar replied:
Sounds good - my only other thought is that the original link was a preprint, whereas the newer link is post publication (and peer review). I don't have a strong preference either way.

rossbar reviewed (commented):
Generally looks good - a couple comments for a few of the links. It would also be nice to switch to using intersphinx for linking the NEPs now that #16450 is in. Thanks @bjnath !

bjnath commented:
With one exception, **every link here was reported broken.**

I ran the checker on a local repo pulled from head, for two reasons:
1. To work on the latest version
2. To be able to verify my fixes worked

Because of 2, I really had no other way to proceed. 

Before changing any link, I **verified** that it was in fact bad. Links in the PR that look OK on the site must be a consequence of working with local files.

After changing each link I verified that it **now worked**. Therefore the new link can't be equivalent to the original. Something got better.

Knowing what I know now, I'd eliminate the false positives by changing one step: When verifying a bad link in the report, I'd check the live page instead of the local page.

The PR wasn't much fun to do and I'm sure not much fun to check. I appreciate reviewers' attention to detail.

bjnath committed:
DOC: Redirect 2 link fixes in PR #16472

* governance.rst link reverted
* ununcs.rst `overridden` link goes where it was meant to


per https://github.com/numpy/numpy/pull/16472#pullrequestreview-424666070

bjnath commented:
2eb904b reverts change in governance.rst, points ufuncs.rst link where it was meant to go.

rossbar commented on doc/source/release/1.5.0-notes.rst:
```diff
@@ -12,11 +12,11 @@ Python 3 compatibility
 Python 3 and Python 2 is done from a single code base. Extensive notes on
 changes can be found at
-`<http://projects.scipy.org/numpy/browser/trunk/doc/Py3K.txt>`_.
+`<https://web.archive.org/web/20100814160313/http://projects.scipy.org/numpy/browser/trunk/doc/Py3K.txt>`_.
```
It turns out this doc is actually included in the `numpy/doc/Py3K.rst.txt`, so the external archive link should be replaced with a reference to the internal doc.

rgommers replied:
For now this is fine I think. That file needs to be moved into the actual docs, rather than being some random text file. And we don't want to have to go back and fix these release notes yet again when doing that (it's likely we forget).

rossbar reviewed (approved):
LGTM - I stumbled across one last link that should be redirected (an archive.org link to a doc that is actually still in the `numpy/doc`) but other than that everything else seems fixed. The NEP/intersphinx issue still exists as well, but as pointed out since that now depends on #16500 , the plan seems to be to not worry about it here.

Thanks for the perseverance @bjnath , I agree that fixing the links is no fun (and quite tedious) but in the end it is an improvement!

rgommers commented:
Merged, thanks a lot @bjnath.

Thanks everyone for reviewing.

rgommers commented:
A comment on review style here: there's 40+ comments for a very straightforward PR that fixes broken links. Keeping on iterating till (a) everything is 100% perfect and (b) includes new improvements on top of fixing broken links is not the best idea. Sometimes perfect is the enemy of good, in particular if we get to round 3 or 4 of changes - the learning moments from the initial feedback has long passed.

There are several better options:
1. Push a fix directly to the branch from which this PR was made for the last nitpicks, then merge.
2. Merge first, then send a follow up PR yourself.
3. Merge first, and ask the contributor if they'd be interested in following up on other improvement opportunities you've seen.

It's still on my todo list to write a reviewers guide, unfortunately I haven't gotten around to it yet. Please take the above as constructive feedback.

rgommers commented:
In case it helps anyone, I use https://hub.github.com/ for making pushing last fixups to PRs easier. The workflow is simple:
```
hub pr checkout 16472   # checks out branch, sets the correct remote
# make local edits as either a new commit, or amend an existing commit
git push
```
That's it, then hit the green button.

rossbar commented:
> A comment on review style here: there's 40+ comments for a very straightforward PR that fixes broken links. 

This is my fault - I'm not 100% confident on what the *best* rST/sphinx syntax is for all the various types of links, so I used the suggestions feature so that I could 1) make the suggestion and 2) explain the reasoning behind it to make it easier to evaluate whether there were any problems with the suggestion.

Sorry for the spam, all. I will try to steer clear of these nitty-gritty rST details in the future!

rgommers commented:
No worries. Thanks @rossbar.

Sphinx is a pain, I'm still not always sure what's "best". The question to ask then is "is this essential to change" - the main thing is that whatever was broken before is now fixed.
